jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9Picture.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright @ 2019 8x8, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.videobridge.cc.vp9
+
+import org.jitsi.nlj.codec.vpx.VpxUtils.Companion.applyExtendedPictureIdDelta
+import org.jitsi.nlj.rtp.VideoRtpPacket
+import org.jitsi.nlj.rtp.codec.vp9.Vp9Packet
+import java.util.Collections
+import kotlin.collections.ArrayList
+
+/**
+ * Groups together some RTP/VP9 fields that refer to a specific incoming VP9
+ * picture, which may consist of multiple frames (of different spatial layers).
+ *
+ * Instances of this class are *NOT* thread safe.
+ *
+ * @author Jonathan Lennox
+ */
+class Vp9Picture(packet: Vp9Packet) {
+    val frames = ArrayList<Vp9Frame?>()
+
+    init {
+        val sid = if (packet.hasLayerIndices) packet.spatialLayerIndex else 0
+
+        setFrameAtSid(Vp9Frame(packet), sid)
+    }
+
+    fun frame(sid: Int) = frames.getOrNull(sid)
+
+    private fun setFrameAtSid(frame: Vp9Frame, sid: Int) {
+        if (sid >= frames.size) {
+            frames.addAll(Collections.nCopies(sid - frames.size, null))
+            frames.add(frame)
```

Comment:
Doesn't the frame still need to be placed at index `sid` ?

Reply:
The previous line expands the `ArrayList` to size `sid` (i.e. with indices `0`...`sid-1`), so `add` will put the new one at index `sid`.

Reply:
Ah ha...I see now.  I think I'd still vote for `frames[sid] = frame` to make that more explicit (and a bit less brittle?), but maybe that's just me.

Reply:
I'm pretty sure `frames[sid] = frame` doesn't work if `frames` is only `sid` elements long (i.e. the maximum addressable element is `frames[sid-1]`.

Reply:
Yes, you're right. It would be `frames[sid - 1] = frame`

Reply:
Right, but I want to set `frames[sid]`.  That's why I use `addAll` to fill in nulls up to the element before `sid`, and then use `add` to set `frames[sid]`.

If `ArrayList` had some sort of `extendAndSet` semantic I'd use it and it'd be clearer, but it doesn't.  Maybe a helper function could be written?

Reply:
Hm, I think I'm confused.  I'm not saying get rid of the `addAll`.  But, either way, you're putting this frame in index `sid - 1`, right?  I agree that it would be clearest to be able to write `frames[sid] = frame`, and `frames[sid - 1] = frame` isn't quite as nice, but it still feels more explicit than just the `frames.add(frame)`, just because I think it conveys that we're adding the frame to the proper index for `sid`, and not just the end.

Reply:
No, I'm putting the frame in at index `sid`, always.  `addAll` fills up the elements *before* `sid`, then `add` sets element `sid`.